### PR TITLE
Add a soft session cap, returning HTTP errors rather than closed connections

### PR DESCRIPTION
**Pull request by eddyashton**
Part of #2521.

Note that this returns errors for _new sessions_, but we don't do anything to discourage/disconnect existing connections. We could develop some scheme to return errors on existing sessions for some kind of fairness, but I don't know if its worth the complexity?

This makes the session exhaustion behaviour far friendlier to clients, and changes what we're able to easily test in `connections_test_cft`.

TODO:
- [x] Add to CHANGELOG
- [X] Check error code. _The issue says 429, but I think 503 (returned in this PR) is more accurate. This error is "the server is overloaded" rather than "you have sent too many requests"._

### CODE REVIEW

File: tests/connections.py [hunk 1]
```diff
@@ -127,6 +147,13 @@ def create_connections_until_exhaustion(target):
         to_create = max_fds - num_fds + 1
         num_fds = create_connections_until_exhaustion(to_create)
 
+        try:
+            create_connections_until_exhaustion(to_create, True)
+        except AllConnectionsCreatedException as e:
+            # This is fine! The soft cap means this test no longer causes reaches the hard cap.
```
Comment (jumaffre) [inline tests/connections.py:153]:
```suggestion
            # This is fine! The soft cap means this test no longer reaches the hard cap.
```

File: src/enclave/rpc_sessions.h [hunk 2]
```diff
@@ -75,26 +114,41 @@ namespace enclave
...
+          max_open_sessions_hard,
           id);
 
         RINGBUFFER_WRITE_MESSAGE(
           tls::tls_stop, to_host, id, std::string("Session refused"));
-        return;
       }
+      else if (sessions.size() >= max_open_sessions_soft)
+      {
+        LOG_INFO_FMT(
```
Comment (jumaffre) [inline src/enclave/rpc_sessions.h:131]: This seems rather verbose?

File: CHANGELOG.md [hunk 3]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Changed
 
 - The curve-id selected for the identity of joining nodes no longer needs to match that of the network (#2525).
+- The per-node session cap behaviour has changed. The `--max-open-sessions` is now a soft cap on the number of sessions. Beyond this, new sessions will receive a HTTP 503 error immediately after completing the TLS handshake. The existing hard cap (where sessions are closed immediately, before the TLS handshake) is still available, under the argument `--max-open-sessions-hard` (#2583).
```
Comment (jumaffre) [inline CHANGELOG.md:20]:
```suggestion
- The per-node session cap behaviour has changed. The `--max-open-sessions` is now a soft cap on the number of sessions. Beyond this, new sessions will receive a HTTP 503 error immediately after completing the TLS handshake. The existing hard cap (where sessions are closed immediately, before the TLS handshake) is still available, under the new argument `--max-open-sessions-hard` (#2583).
```

**Review by jumaffre — APPROVED** [review]

**Comment by achamayou** [comment]
@eddyashton doing this only for new sessions seems fine. Shutting down existing sessions to give priority to fresh ones doesn't sound great, a single malicious actor could easily DoS the service, whereas right now at least existing connections would continue.

File: src/enclave/rpc_sessions.h [hunk 4]
```diff
@@ -38,6 +39,43 @@ namespace enclave
...
+
+        if (get_status() == Status::ready)
+        {
+          // Send HTTP response describing soft session limit
+          auto http_response = http::Response(HTTP_STATUS_SERVICE_UNAVAILABLE);
+          http_response.set_header(
+            http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+          const auto response_body = fmt::format(
+            "Service is currently busy and unable to serve new connections");
+          http_response.set_body(
```
Comment (achamayou) [inline src/enclave/rpc_sessions.h:64]: Seems like we could and perhaps even should have a set_body(const std::string&)?

File: src/enclave/rpc_sessions.h (hunk 2 above)
Comment (eddyashton) [reply src/enclave/rpc_sessions.h:131]: I have this at `INFO` to match the existing logging when the cap is reached. I think it shouldn't happen under normal operation, and whenever it _does_ happen the operator should be informed. Especially with the other change from #2521, where the operator is likely to be using a separate interface and not seeing this cap, its important that they have some signal that the service is overloaded. Maybe with #2522 they have another way to find this, but I think the log is still valuable.

**Review by achamayou — APPROVED** [review]

File: src/enclave/rpc_sessions.h (hunk 4 above)
Comment (eddyashton) [reply src/enclave/rpc_sessions.h:64]: I don't think this is worth adding. Accessing the raw `http::Response` like this is pretty rare, and its interface surface is currently pretty small. I don't think this conversion is too onerous here, or repeated elsewhere.

**Pull request merged**